crates/commitlog/src/payload.rs

```diff
@@ -53,3 +81,383 @@ impl<const N: usize> Decoder for ArrayDecoder<N> {
+        buf.put_slice(self.reducer_name.as_bytes());
+        buf.put_slice(&self.reducer_args);
+    }
+
+    pub fn decode<'a, R: BufReader<'a>>(reader: &mut R) -> Result<Self, DecodeError> {
+        let len = reader.get_u32()?;
+        let slen = reader.get_u8()?;
+        let reducer_name = {
+            let bytes = reader.get_slice(slen as usize)?;
+            Varchar::from_str(std::str::from_utf8(bytes)?)
+                .expect("slice len cannot be > 255")
+                .into()
+        };
+        let reducer_args = reader.get_slice(len as usize - /* slen */ 1 - slen as usize)?.into();
+
+        Ok(Self {
+            reducer_name,
+            reducer_args,
+        })
+    }
+}
+
+/// The outputs of a transaction.
+///
+/// The only currently possible output of a transaction is a string
+/// representation of an `Err` return value of a reducer.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Outputs {
+    // TODO: We may want a `Cow`-backed variant of `Varchar` for this.
+    pub reducer_output: Arc<Varchar<255>>,
+}
+
+impl Outputs {
+    pub fn encode(&self, buf: &mut impl BufWriter) {
+        let slen = self.reducer_output.len() as u8;
+        buf.put_u8(slen);
+        buf.put_slice(self.reducer_output.as_bytes());
+    }
+
+    pub fn decode<'a, R: BufReader<'a>>(reader: &mut R) -> Result<Self, DecodeError> {
+        let slen = reader.get_u8()?;
+        let reducer_output = {
+            let bytes = reader.get_slice(slen as usize)?;
+            Varchar::from_str(std::str::from_utf8(bytes)?).unwrap().into()
+        };
+
+        Ok(Self { reducer_output })
+    }
+}
+
+/// Mutations of the data store performed by a transaction.
+///
+/// All operations are supposed to be **mutually exclusive**.
+///
+/// Even though not enforced, each kind of mutation should not contain duplicate
+/// [`TableId`]s.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Mutations {
+    /// Rows inserted.
+    pub inserts: Box<[Ops]>,
+    /// Rows deleted.
+    pub deletes: Box<[Ops]>,
+    /// Truncated tables.
+    pub truncates: Box<[TableId]>,
+}
+
+impl Mutations {
+    /// Encode `self` in the canonical format to the given buffer.
+    pub fn encode(&self, buf: &mut impl BufWriter) {
+        encode_varint(self.inserts.len(), buf);
+        for ops in self.inserts.iter() {
+            ops.encode(buf);
+        }
+        encode_varint(self.deletes.len(), buf);
+        for ops in self.deletes.iter() {
+            ops.encode(buf);
+        }
+        encode_varint(self.truncates.len(), buf);
+        for TableId(table_id) in self.truncates.iter() {
+            buf.put_u32(*table_id);
+        }
+    }
+
+    /// Decode `Self` from the given buffer.
+    ///
+    /// A [`RowDecoder`] is supplied to allow to determine the row types of
+    /// individual [`Ops::rowdata`] rows.
+    pub fn decode<'a, D, R>(row_decoder: &D, reader: &mut R) -> Result<Self, DecodeError>
```

Comment:
Turns out this won’t work in practice because ops can depend on schema modifications within the same transaction. Which means that the row decoder would need to update its internal state after each decoder row.

Reply:
I thought we decided that in the case of a schema mutation, all rows in the transactions would be converted to the target schema before saving:

https://github.com/clockworklabs/SpacetimeDBPrivate/pull/718/files#diff-84260e38cfce8e0fba4beb7475cba661892f4a1a66b669db5d49830cc3d43c11R388

Reply:
This has been re-implemented using a visitor pattern, which can be optimized further to not allocate the `Txdata` struct at all when replaying (aka folding) the history.